force-app/main/default/classes/SOQL.cls

```diff
@@ -549,7 +570,39 @@ public virtual inherited sharing class SOQL implements Queryable {
     }
 
     public Map<Id, SObject> toMap() {
-        return new Map<Id, SObject>(toList());
+        Map<Id, SObject> idToSObject = (Map<Id, SObject>) Type.forName('Map<Id, ' + ofObject  + ' >').newInstance();
+        idToSObject.putAll(toList());
+        return idToSObject;
+    }
+
+    public Map<String, SObject> toMap(SObjectField keyField) {
+        with(keyField);
+
+        Map<String, SObject> cutomKeyToRecord = (Map<String, SObject>) Type.forName('Map<String, ' + ofObject  + ' >').newInstance();
+
+        for (SObject record : toList()) {
+            cutomKeyToRecord.put(record.get(keyField) + '', record);
+        }
+
+        return cutomKeyToRecord;
+    }
+
+    public Map<String, List<SObject>> toAggregatedMap(SObjectField keyField) {
+        with(keyField);
+
+        Map<String, List<SObject>> cutomKeyToRecords = (Map<String, List<SObject>>) Type.forName('Map<String, List<' + ofObject  + ' >>').newInstance();
+
+        for (SObject record : toList()) {
+            String customKey = record.get(keyField) + '';
```

Comment:
`String.valueOf` won't have the same result? It will be more clear what is done here

Reply:
Yes, I had it, but when key is null - `String.valueOf(null)`. It will not work. `null + ''` works. Apex <3